rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,1010 @@
+which can then be verified easily on the server.
+
+A rough outline of the joining procedure:
+
+```mermaid
+sequenceDiagram
+	participant keystore as Local Keystore
+	participant bot as Bot
+	participant auth as Auth Server
+
+	opt first join & has initial join token
+		bot->>+keystore: Generate new keypair
+		keystore-->>-bot: Public key
+
+		bot->>auth: Bind public key with token
+		auth->>auth: Verify token<br>& record public key
+	end
+
+	bot->>auth: Request joining challenge
+	auth-->>bot: Sends joining challenge
+	bot->>keystore: Request signed document
+	keystore-->>bot: Signed challenge document
+	bot->>auth: Signed challenge document<br>& previous join state
+	auth->>auth: Validate signed document<br>against bound public key
+
+	opt no valid client certificate
+		auth->>auth: Validate join state document
+		auth->>auth: decrement rejoin counter
+	end
+
+	auth-->>bot: Signed TLS certificates<br>& new join state
+```
+
+To avoid use of traditional renewable certificates, this takes advantage of
+recent support for optionally authenticated joining added as part of
+[RFD 0162](./0162-machine-id-token-join-method-bot-instance.md). If a bot still
+has a valid client certificate from a previous authentication attempt, it uses
+it to open an mTLS session.
+
+When Auth validates the join attempt, clients that presented an existing valid
+identity are considered to be requesting a refresh rather than rejoining,
+leaving the rejoin counter untouched. Clients that do not present a valid client
+certificate are considered to be rejoining and the token associated with this
+public key must have `.status.bound_keypair.remaining_rejoins` >= 1.
+
+#### Client-Side Changes in `tbot`
+
+Bots should be informed of various status metrics, including number of remaining
+rejoins and whether or not a keypair rotation has been requested. There's a few
+methods by which we could inform bots of their remaining rejoins:
+
+1. (Recommended) Heartbeats: bots submit heartbeats at startup and on a regular
+   interval. It would be trivial to include a remaining rejoin counter in the
+   (currently empty) heartbeat response.
+
+2. Certificate field: we could include the number of remaining rejoins in a
+   certificate field.
+
+3. New RPC: we could add a new RPC for bots to fetch this, alongside any other
+   potentially useful information.
+
+4. We could grant bots permission to view their own join tokens. There is
+   precedent here as bots can view e.g. their own roles without explicitly
+   having RBAC permissions to do so.
+
+The remaining rejoin counter should then be exposed as a Prometheus metric to
+allow for alerting if a bot drops below some threshold.
+
+Importantly, this is a potentially lagging indicator. The design allows for the
+rejoin counter to be decreased (to zero) at any time, so a rejoin attempt may
+still fail at any time. This should be acceptable since it can also be increased
+after the fact to restore access if desired.
```

Comment:
Given that we need to handle this failure-mode anyway, is there much benefit to pushing the rejoin counter and whether a key rotation is required to the bot proactively? (as in the heartbeating suggestion above).

Instead, we could just return the rejoin counter on each renewal, and return a "you need to rotate your public key" response when trying to renew after `rotate_on_next_renewal` has been set.

Reply:
Following up on our discussion, I think the rejoin counter could be included in the join state document, and I'll at least move the `rotate_on_next_renewal` flag here. I think the forced-rotation-via-error flow has some merits but for simplicity of implementation I have a mild preference to let clients initiate the rotation process.